File: android/takanawa-android/src/main/kotlin/ai/yetanother/takanawa/Takanawa.kt
package ai.yetanother.takanawa

object Takanawa {
    @JvmStatic
    @JvmOverloads
    fun init(maxIo: Int = 0) {
        require(maxIo >= 0) { "maxIo must be greater than or equal to 0" }
        checkStatus(NativeBridge.globalInit(maxIo))
    }

    @JvmStatic
    fun setMaxIo(maxIo: Int) {
        require(maxIo >= 0) { "maxIo must be greater than or equal to 0" }
        checkStatus(NativeBridge.globalSetMaxIo(maxIo))
    }

    @JvmStatic
    fun shutdown() {
        checkStatus(NativeBridge.globalShutdown())
    }
}

File: packages/takanawa-csharp/src/YetAnotherAI.Takanawa/Takanawa.cs
using System;

namespace YetAnotherAI.Takanawa
{
    public static class Takanawa
    {
        public static void Init(int maxIo = 0)
        {
            if (maxIo < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(maxIo), maxIo, "maxIo must be greater than or equal to 0");
            }

            var config = new NativeMethods.TknwGlobalConfig
            {
                AbiVersion = NativeMethods.AbiVersion,
                StructSize = NativeConversions.SizeOf<NativeMethods.TknwGlobalConfig>(),
                MaxIo = NativeConversions.ToUIntPtr(maxIo, nameof(maxIo)),
            };
            CheckStatus(NativeMethods.GlobalInit(ref config));
        }

        public static void SetMaxIo(int maxIo)
        {
            if (maxIo < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(maxIo), maxIo, "maxIo must be greater than or equal to 0");
            }

            CheckStatus(NativeMethods.GlobalSetMaxIo(NativeConversions.ToUIntPtr(maxIo, nameof(maxIo))));
        }

        public static void Shutdown()
        {
            CheckStatus(NativeMethods.GlobalShutdown());
        }

        internal static void CheckStatus(TakanawaStatus status, IntPtr handle = default)
        {
            var normalized = NativeConversions.NormalizeStatus(status);
            if (normalized == TakanawaStatus.Ok)
            {
                return;
            }

            string? nativeMessage = null;
            if (handle != IntPtr.Zero)
            {
                nativeMessage = NativeBuffers.ReadLastError(handle);
            }

            throw new TakanawaException(normalized, nativeMessage);
        }
    }
}
